### DISCUSSION TITLE
Seeing some strange behavior in profiling

### DISCUSSION THREAD

**Discussion by cooljacob204** [jeremyevans/sequel#2207] [General]
Hey all,

I'm seeing some strange behavior in some of our threads with them spending a lot of time this sleep call: https://github.com/jeremyevans/sequel/blob/master/lib/sequel/connection_pool/threaded.rb#L297

This is a job within sidekiq.

![image](https://github.com/user-attachments/assets/3a78d9f5-db31-447d-bad7-5e99d3f4f73d)

My first though is that we have a cpu issue. But in the profiling at least it doesn't seem to be indicate that sleep is using the most across all threads in various jobs. We do have some high CPU usage in some workers but not all of them.

~~Wouldn't high CPU usage make threads want to swap faster therefor spend less time sleeping?~~ 
Actually maybe that makes sense. It's sleeping longer because another thread grabbed context causing this one to stay sleeping. Does that sound right? It still doesn't explain to me though why most of the profiling time is sleeping.

Any idea what could cause our threads to spend a lot of time in the sleep method?

**Comment by jeremyevans** [comment]
This is a `sleep(0)` call, used to pass control to another thread.  If you are using a sampling profiler, this is likely due to the sampling.  Can you try a tracing profiler and see how much time is actually spent in the `sleep(0)` call?  I would guess it is insignificant.

If this is causing an actual issue, I'm OK with making the `sleep(0)` optional.  It appears to be added to work around issues when using the async library.

Also, you may want to try using the timed_queue connection pool (`SEQUEL_DEFAULT_CONNECTION_POOL=timed_queue` environment variable, or `:pool_class=>:timed_queue` database option).  That is the default connection pool in Ruby 3.4+, but should work in Ruby 3.2+ (it may become the default connection pool in Ruby 3.2 and 3.3).

**Reply by cooljacob204** [reply]
So I'm using Datadog which I believe uses sampling. They do have a trace service but it doesn't give me a lot of details. Here is an example of a trace. Which you can see it spends very little time in PG but a lot of time in the Sequel library. (green is PG, dark pink is sequel)

![image](https://github.com/user-attachments/assets/136ebf86-cd95-4efb-a8a5-29436b21223d)

My instinct is that it's a resource contention issue but these two things point to Sequel using the resources somehow, which I know is probably not right but hoping to better understand what could cause it to appear this way.

I'm also going to look into changing to `timed_queue`. So far I haven't found much documentation on what it does though.

**Reply by jeremyevans** [reply]
It isn't unusual for the majority of the time to be spent in Sequel, unless the queries you are running are quite heavy on the database.

The timed_queue connection pool operates similar to the default threaded connection pool, but it uses the `Queue#pop` `:timeout` option added in Ruby 3.2.  In most cases, it is a drop-in replacement.  The timed_queue connection pool doesn't have a `sleep(0)` call, because the `sleep(0)` is to work around an issue with ConditionVariables, and it does not use a ConditionVariable (the threaded connection pool uses a ConditionVariable).

If you aren't already using `sequel_pg`, you may want to consider it, as it can significantly reduce the time needed for Sequel to handle typecasting of PostgreSQL result sets.

**Reply by cooljacob204** [reply]
Thanks Jeremy! You have given me a lot to go off of. Changing to timed_queue made our traces a lot better and we can now see where our worker is really spending its time.

Also thanks for creating and maintaining my favorite gem!

**Reply by cooljacob204** [reply]
So it looks like this fixed our problem completely. I'm sorta baffled, I don't really understand why this would fix it. 

Our time spend processing the problematic job:
![image](https://github.com/user-attachments/assets/ab9ba964-869c-4d1d-884b-f9d84b39ff5e)
Our 1 hour queue latency where the job lived:
![image](https://github.com/user-attachments/assets/c50d4c12-1015-4c4c-a27a-4a81d886aa55)
New profile:
![image](https://github.com/user-attachments/assets/8ef13fa8-c341-4bdd-a24f-7a1a359d37f8)

Going to verify this by going back to the old pool setting and see if it spikes.

**Reply by jeremyevans** [reply]
Please do.  I had a hunch that `timed_queue` was better, but I had no idea the difference was that dramatic.

**Reply by cooljacob204** [reply]
Yeah, turning it back off dramatically increased the amount of time my jobs took, I turned it off at 17 utc and back on at 18 utc.

Our sidekiq queue latency:
![image](https://github.com/user-attachments/assets/69fa6379-4558-4f75-8dc7-50d051aa4bc6)

Time spent performing the job:
![image](https://github.com/user-attachments/assets/6bb24849-1c36-4d89-bae4-40af695d5e33)

For additional context we are on Ruby 3.3.3, Rails 7.1.3.2, Sidekiq 6.5.12, Sequel 5.77.0

**Reply by cooljacob204** [reply]
Our workers have a db_pool size of 10 and a sidekiq concurrency of 10.

Even when I reduced the sidekiq concurrency to 5 we still saw this issue.

**Reply by jeremyevans** [reply]
Thank you very much for that information.  I'll accelerate my plans to make timed_queue the default connection pool on Ruby 3.2+.
